InfiniTAM/Engine/Kinect2Engine.cpp

```diff
@@ -87,43 +99,136 @@ Kinect2Engine::~Kinect2Engine()
 
 void Kinect2Engine::getImages(ITMUChar4Image *rgbImage, ITMShortImage *rawDepthImage)
 {
-	//Vector4u *rgb = rgbImage->GetData(MEMORYDEVICE_CPU);
-	//if (colorAvailable)
-	//{
-	//}
-	//else memset(rgb, 0, rgbImage->dataSize * sizeof(Vector4u));
-
-	//float *depth = out->depth->GetData(MEMORYDEVICE_CPU);
-	//if (depthAvailable)
-	//{
-	//	IDepthFrame* pDepthFrame = NULL;
-	//	UINT16 *pBuffer = NULL;
-	//	UINT nBufferSize = 0;
-
-	//	HRESULT hr = data->depthFrameReader->AcquireLatestFrame(&pDepthFrame);
-
-	//	if (SUCCEEDED(hr))
-	//	{
-	//		if (SUCCEEDED(hr))
-	//			hr = pDepthFrame->AccessUnderlyingBuffer(&nBufferSize, &pBuffer);
-
-	//		if (SUCCEEDED(hr))
-	//		{
-	//			for (int i = 0; i < imageSize_d.x * imageSize_d.y; i++)
-	//			{
-	//				ushort depthPix = pBuffer[i];
-	//				depth[i] = depthPix == 0 ? -1.0f : (float)depthPix / 1000.0f;
-	//			}
-	//		}
-	//	}
-
-	//	SafeRelease(pDepthFrame);
-	//}
-	//else memset(depth, 0, out->depth->dataSize * sizeof(short));
-
-	//out->inputImageType = ITMView::InfiniTAM_FLOAT_DEPTH_IMAGE;
-
-	return /*true*/;
+	Vector4u *rgb = rgbImage->GetData(MEMORYDEVICE_CPU);
+
+	if (colorAvailable)
+	{
+		IFrameDescription* pFrameDescription = NULL;
+		int nWidth = 0;
+		int nHeight = 0;
+		IColorFrame* pColorFrame = NULL;
+		RGBQUAD *pBuffer = NULL;
+		UINT nBufferSize = 0;
+
+		HRESULT hr = data->colorFrameReader->AcquireLatestFrame(&pColorFrame);
+
+		// Get frame descriptor to get the width and the height of the image frame.
+		if (SUCCEEDED(hr))
+			hr = pColorFrame->get_FrameDescription(&pFrameDescription);
+
+		if (SUCCEEDED(hr))
+			hr = pFrameDescription->get_Width(&nWidth);
+
+		if (SUCCEEDED(hr))
+			hr = pFrameDescription->get_Height(&nHeight);
+
+		SafeRelease(pFrameDescription);
+
+		RGBQUAD* m_pColorRGBX = new RGBQUAD[cColorWidth * cColorHeight];
+
+		if (SUCCEEDED(hr))
+		{
+			// Must first determine the color image format
+			ColorImageFormat imageFormat = ColorImageFormat_None;
+			hr = pColorFrame->get_RawColorImageFormat(&imageFormat);
+
+			if (imageFormat == ColorImageFormat_Bgra)
+			{
+				hr = pColorFrame->AccessRawUnderlyingBuffer(&nBufferSize, reinterpret_cast<BYTE**>(&pBuffer));
+			}
+			else if (m_pColorRGBX)
```

Comment:
The test you're doing here suggests that you may be thinking that new will return a null pointer if it fails; instead, it will normally throw an exception on failure unless you use std::nothrow:

http://en.cppreference.com/w/cpp/memory/new/nothrow